chiron/mcmc.py

```diff
@@ -112,103 +175,802 @@ def run(
+            Whether to update the stepsize of the move. Default is True.
+        update_stepsize_frequency : int, optional
+            Frequency of updating the stepsize of the move. Default is 100.
+        Returns
+        -------
+        None
+        """
+        super().__init__(
+            nr_of_moves=nr_of_moves,
+            reporter=reporter,
+            report_frequency=report_frequency,
+            update_stepsize=update_stepsize,
+            update_stepsize_frequency=update_stepsize_frequency,
+            method="metropolis",
+        )
+        self.volume_max_scale = volume_max_scale
+
+    def _report(
+        self,
+        step: int,
+        iteration: int,
+        elapsed_step: int,
+        acceptance_probability: float,
+        sampler_state: SamplerState,
+        thermodynamic_state: ThermodynamicState,
+        nbr_list: Optional[PairsBase] = None,
+    ):
+        """
+
+        Parameters
+        ----------
+        step : int
+            The current step of the simulation move.
+        iteration : int
+            The current iteration of the move sequence (i.e., how many times has this been called thus far).
+        elapsed_step : int
+            The total number of steps that have been taken in the simulation move. step+ nr_moves*iteration
+        acceptance_probability : float
+            The acceptance probability of the move.
+        sampler_state : SamplerState
+            The sampler state of the system.
+        thermodynamic_state : ThermodynamicState
+            The thermodynamic state of the system.
+        nbr_list : Optional[PairBase]=None
+            The neighbor list or pair list for evaluating interactions in the system, default None
+        """
+
+        potential = thermodynamic_state.potential.compute_energy(
+            sampler_state.x0, nbr_list
+        )
+        volume = (
+            sampler_state.box_vectors[0][0]
+            * sampler_state.box_vectors[1][1]
+            * sampler_state.box_vectors[2][2]
+        )
+        self.reporter.report(
+            {
+                "step": step,
+                "iteration": iteration,
+                "elapsed_step": elapsed_step,
+                "potential_energy": potential,
+                "volume": volume,
+                "box_vectors": sampler_state.box_vectors,
+                "max_volume_scale": self.volume_max_scale,
+                "acceptance_probability": acceptance_probability,
+            }
+        )
+
+    def _update_stepsize(self):
+        """
+        Update the volume_max_scale parameter to ensure our acceptance probability is within the range of 0.25 to 0.75.
+        The maximum volume_max_scale will be capped at 0.3.
+        """
+        acceptance_ratio = self.n_accepted / self.n_proposed
+        if acceptance_ratio < 0.25:
+            self.volume_max_scale /= 1.1
+        elif acceptance_ratio > 0.75:
+            self.volume_max_scale = min(self.volume_max_scale * 1.1, 0.3)
+
```

Comment:
Is there a way to refactor this into one of the base classes? This seems to be super helpful for general MC moves. But, it might also be specific enough so that abstracting it becomes difficult, then we can just leave it.

Reply:
I defined it in the base class as an abstract method. I don’t think this will be general enough to put the algorithm in the base class (this is pretty specific to simple displacement and volume changes moves…would not apply to a insertion/deletion move).  We could make this a class as I suggested in an issue #26 , that we pass to the init function to set up the protocol (and nothing is adjusted if not set).

Reply:
In #26, we discussed creating something like an `Autotune` object that handles auto-tuning behavior. That could provide an API for auto-tuning operations that adapt parameters like these, and could eventually implement asymptotically consistent approaches to make sure that we don't cause issues with sampling.

Reply:
Yes, I think creating an Autotune class will be a good idea, especially for anything more advanced.  I think that will be easy to implement in a separate PR so we can better discuss the API we want.